File: C_Tools/PcieCamTool/PcieCamTool.c
#include "stdio.h"
#include "sys/io.h"
#include "string.h"
#include "stdbool.h"

#define MaxBusNum                       256
#define MaxDevNum                       32
#define MaxFunNum                       8

//
// Declare I/O Ports used to perform PCI Confguration Cycles
//
#define PCI_CONFIGURATION_ADDRESS_PORT  0xCF8
#define PCI_CONFIGURATION_DATA_PORT     0xCFC


#define PCI_IO_ADDRESS(Bus, Dev, Fun, Reg) 0x80000000+(Bus<<16)+(Dev<<11)+(Fun<<8)+Reg


/**
  Print usage.
**/
void
PrintUsage (
  void
  )
{
  printf("PcieCamTool:  usage\n");
  printf("  PcieCamTool -h\n");
  printf("  PcieCamTool -x\n");
  printf("  PcieCamTool -s Bus:Dev.Fun\n");
  printf("  PcieCamTool\n");
  printf("Parameter:\n");
  printf("  -h -help: show instructions\n");
  printf("  -s: specify a pcie device with the format of Bus:Dev.Fun.\n");
  printf("  -x: show pcie configuration space\n");
}

bool
IsDecimalDigitCharacter (
  char                    Char
) {
  return (bool) (Char >= '0' && Char <= '9');
}

char
Char16ToUpper (
  char                    Char
  )
{
  if (Char >= 'a' && Char <= 'z') {
    return (char) (Char - ('a' - 'A'));
  }

  return Char;
}

int
HexStrToUint16 (
  char              FirstChara,
  char              SecondChara
  )
{
  unsigned int      Data;

  if(FirstChara == SecondChara) {
    if (IsDecimalDigitCharacter(FirstChara)) {
      Data = FirstChara - '0';
    } else {
      if ((Char16ToUpper(SecondChara) - 'A') > 5) return -1;
      Data = Char16ToUpper(FirstChara) - 'A' + 10;
    }
  } else {
    if (IsDecimalDigitCharacter(FirstChara)) {
      Data = FirstChara - L'0';
    } else {
      if ((Char16ToUpper(FirstChara) - 'A') > 5) return -1;
      Data = Char16ToUpper(FirstChara) - 'A' + 10;
    }

    if (IsDecimalDigitCharacter(SecondChara)) {
      Data = (Data << 4) + (SecondChara - '0');
    } else {
      if ((Char16ToUpper(SecondChara) - 'A') > 5) return -1;
      Data = (Data << 4) + (Char16ToUpper(SecondChara) - 'A' + 10);
    }
  }

  return Data;
}

/**
  Print input parameters.
**/
int
ParseParameters (
  char                *InputStr,
  int                 *Bus,
  int                 *Dev,
  int                 *Fun 
  )
{
  unsigned short      Index;
  unsigned short      SemicolonPosition;
  unsigned short      DotPisition;
  unsigned short      InputStrLen;

  if (InputStr == NULL) {
    return -1;
  }

  for (Index = 0; InputStr[Index] != '\0'; Index ++) {
    if ((InputStr[Index] == ' ') || (InputStr[Index] == '\t')) return -1;
    if (InputStr[Index] == ':') SemicolonPosition = Index;
    if (InputStr[Index] == '.') DotPisition       = Index;
  }

  InputStrLen = Index;

  if ((SemicolonPosition > 2) || (DotPisition - SemicolonPosition > 3) || (InputStrLen - DotPisition > 2)) {
    return -1; 
  }

  *Bus = HexStrToUint16(InputStr[0], InputStr[SemicolonPosition - 1]);
  *Dev = HexStrToUint16(InputStr[SemicolonPosition + 1], InputStr[DotPisition - 1]);
  *Fun = HexStrToUint16(InputStr[DotPisition + 1], InputStr[InputStrLen - 1]);

  if (*Bus == -1 || *Dev == -1 || *Fun == -1) return -1;

  printf ("InputStrLen = %d, Bus = %x, Dev = %x, Fun = %x\n", InputStrLen, *Bus, *Dev, *Fun);

  return 0;
}

void
ShowPciConfigSpace (
  unsigned long    PciIoAddr
) {
  unsigned short   Index;
  unsigned char    Data;

  // io access can only read the first 256 bytes
  for (Index = 0; Index <= 0xFF; Index ++) {
    outl ((PciIoAddr + Index), PCI_CONFIGURATION_ADDRESS_PORT);
    Data = inb (PCI_CONFIGURATION_DATA_PORT + ((PciIoAddr + Index) & 3));

    printf("%02x ", Data);
    if ((Index + 1) % 16 == 0 ) printf("\n");
  }
}

int EnableIoAccess (
  void
) {
  // get permission to access io ports
  // for i386 system, there are 2 functions iopl() and ioperm() that can set io port input/output permissions.
  // the differences is that  ioperm() supports only io ports in the range of0-0x3ff; 
  // but iopl() supports 65536 io ports.

  // int ioperm(unsigned long from, unsigned long num, int turn_on); => can not use ioperm here
  // int iopl(int level);

  if (iopl(3) != 0) {
    return -1;
  }
  return 0;
}

int DisableIoAccess (
  void
) {
  if (iopl(0) != 0) {
    printf("Disable io port accessing permission failed.\n");
    return -1;
  }
  return 0;
}

int main (
  int           Argc,
  char          *Argv[]
) {
  unsigned int  BusIndex;
  unsigned int  DevIndex;
  unsigned int  FunIndex;
  int           Bus;
  int           Dev;
  int           Fun;
  unsigned long PciIoAddr;
  unsigned int  VId;
  unsigned int  DId;
  int           ReturnValue;
  unsigned int  Index;
  bool          ShowRegs;

  PciIoAddr     = 0;
  VId           = 0;
  DId           = 0;
  Index         = 0;
  ShowRegs      = false;

  printf("Scanning Pcie devices by CAM(IO port) method:\n\n");

  // Parse input parameters.
  for (Index = 0; Index < Argc; Index ++) {
    if (strcmp(Argv[Index], "-h") == 0 || strcmp(Argv[Index], "-help") == 0) {
      PrintUsage();
      return 0;
    }
  }

  for (Index = 0; Index < Argc; Index ++) {
    if (strcmp(Argv[Index], "-x") == 0 || strcmp(Argv[Index], "-X") == 0) {
      ShowRegs = true;
    }
  }

  if (EnableIoAccess() != 0) {
    printf("Get io port accessing permission failed.\n");
    DisableIoAccess();
    return -1;
  }

  for (Index = 0; Index < Argc - 1; Index ++) {
    if (strcmp(Argv[Index], "-s") == 0 || strcmp(Argv[Index], "-S") == 0) {
      ReturnValue = ParseParameters(Argv[Index + 1], &Bus, &Dev, &Fun);
      if (ReturnValue == -1) {
        printf("Invalid Pci address.\n");
        DisableIoAccess();
        return -1;
      }

      PciIoAddr = PCI_IO_ADDRESS(Bus, Dev, Fun, 0);
      outl (PciIoAddr, PCI_CONFIGURATION_ADDRESS_PORT);
      VId = inw(PCI_CONFIGURATION_DATA_PORT);
      if (VId == 0xffff) {
        printf("Invalid Pci address.\n");
        DisableIoAccess();
        return -1;
      }

      DId = inw(PCI_CONFIGURATION_DATA_PORT + 2);
      printf("%02x:%02x.%x  VId = 0x%04x, DId = 0x%04x\n", Bus, Dev, Fun, VId, DId);
      if (ShowRegs) {
        ShowPciConfigSpace(PciIoAddr);
        printf("\n");
      }
    DisableIoAccess();
    return 0;
    }
  }

  for (BusIndex = 0; BusIndex < MaxBusNum; BusIndex ++) {
    for (DevIndex = 0; DevIndex < MaxDevNum; DevIndex ++) {
      for (FunIndex = 0; FunIndex < MaxFunNum; FunIndex ++) {

        PciIoAddr = PCI_IO_ADDRESS(BusIndex, DevIndex, FunIndex, 0);
        outl (PciIoAddr, PCI_CONFIGURATION_ADDRESS_PORT);
        VId = inw(PCI_CONFIGURATION_DATA_PORT);
        if ((VId == 0xffff) && (FunIndex == 0)) {
          break;
        } else if (VId == 0xffff) {
          continue;
        }

        DId = inw(PCI_CONFIGURATION_DATA_PORT + 2); 
        printf ("%02x:%02x.%x  VId = 0x%04x, DId = 0x%04x\n", BusIndex, DevIndex, FunIndex, VId, DId);
        if (ShowRegs) {
          ShowPciConfigSpace(PciIoAddr);
          printf ("\n");
        }
      }
    }
  }

  DisableIoAccess();

  return 0;
}

File: Python_Tools/PcieCamTool/PcieCamTool.py
import os
import sys
import getopt
import portio

MaxBusNum                         =   256
MaxDevNum                         =   32
MaxFuncNum                        =   8

PCI_CONFIGURATION_ADDRESS_PORT    = 0xCF8
PCI_CONFIGURATION_DATA_PORT       = 0xCFC

def PCI_IO_ADDRESS (Bus, Dev, Func, Reg):
    return 0x80000000+(Bus<<16)+(Dev<<11)+(Fun<<8)+Reg

def EnableIoAccess ():
    return portio.iopl(3)

def DisableIoAccess ():
    return portio.iopl(0)

def ShowPciConfigSpace (Addr):
    for Index in range(0, 0x1000):
        if Index == 0x100 and MmioRead(Addr + Index, 4) == [0, 0, 0, 0]:
            break
        Data = MmioRead(Addr + Index, 1)
        print ("%02x " %(Data[0]), end="")
        if (Index + 1) % 16 == 0:
            print("")
    
def HelpMessage ():
    print("PcieEcamTool:  usage")
    print("  PcieEcamTool -h")
    print("  PcieEcamTool -x")
    print("  PcieEcamTool -s Bus:Dev.Func")
    print("  PcieEcamTool")
    print("Parameter:")
    print("  -h -help: show instructions")
    print("  -s: specify a pcie device with the format of Bus:Dev.Func")
    print("  -x: show pcie configuration space")

def HexStringToInt (str):
    Val = str.lower()
    return int(Val, 16)

if __name__ == "__main__":

    # check for root privileges
    if os.getuid():
        print('This program needs to be root! Exiting.')
        sys.exit()

    InputPcieDev = ""
    ShowRegs = False
    print("\nPython tool for doing Pcie devices by CAM(CPU IO):\n")

    # Parse input parameters.
    opts,args = getopt.getopt(sys.argv[1:], "hHxXs:S:", )
    for opt_name,opt_value in opts:
        if opt_name in ("-h", "-H"):
            HelpMessage()
            sys.exit()
        elif opt_name in ("-x", "-X"):
            ShowRegs = True
        elif opt_name in ("-s", "-S"):
            InputPcieDev = opt_value

    ReturnVal = EnableIoAccess()
    if ReturnVal != 0:
        print('Get io port accessing permission failed.')
        DisableIoAccess()
        sys.exit()

    if InputPcieDev != "":
        parts = InputPcieDev.split(":")
        Bus = HexStringToInt(parts[0])
        subparts = parts[1].split(".")
        Dev = HexStringToInt(subparts[0])
        Func = int(subparts[1])
        # print (Bus, Dev, Func)
        if Bus < 0 or Bus >= MaxBusNum or Dev < 0 or Dev >= MaxDevNum or Func < 0 or Func >= MaxFuncNum:
            print("Input parameter for Bus:Dev.Func is wrong.\n")
            HelpMessage()
            sys.exit()
        else:
            PcieIoAddr = PCI_IO_ADDRESS(Bus, Dev, Func, 0)
            VId =
            if VId[0] == 0xff and VId[1] == 0xff:
                print("Invalid Pci address")
                sys.exit()
            DId = 
            print("%02x:%02x.%x  VId = 0x%02x%02x, DId = 0x%02x%02x" %(Bus, Dev, Func, VId[1], VId[0], DId[1], DId[0]))
            if ShowRegs == True:
                ShowPciConfigSpace(PcieIoAddr)
                print("\n")
    else:
        for Bus in range(0, MaxBusNum):
            for Dev in range(0, MaxDevNum):
                for Func in range(0, MaxFuncNum):
                    PcieIoAddr = PCI_IO_ADDRESS(Bus, Dev, Func, 0)
                    VId = 
                    if VId[0] == 0xff and VId[1] == 0xff:
                        continue
                    DId = 
                    print("%02x:%02x.%x  VId = 0x%02x%02x, DId = 0x%02x%02x" %(Bus, Dev, Func, VId[1], VId[0], DId[1], DId[0]))
                    if ShowRegs == True:
                        ShowPciConfigSpace(PcieIoAddr)
                        print("\n")